### DISCUSSION TITLE
How to use global data in eleventyComputed?

### DISCUSSION THREAD

**Discussion by opexchanger** [11ty/eleventy#1828] [Q&A]
I might not understand correctly how to use the eleventyComputed property.
I created a layout to be used by all my pages. The regular data flows alright, and I am also able to set some variables from the partials to the layout via front-matter:
```
---
layout: layout.html
page: 'home'
---
```
My issue starts with the <title> of the pages, inside the head of this layout. I cannot simply pass the value as a string via front-matter, because each page gets it's title from the CMS.
So I tried passing the data value from global data with no success:

```
eleventyComputed:
  pageTitle: "{{ home.pageTitle.pt }}"
---
```
This literally outputs: {{ home.pageTitle.pt }} as a string; and
```
eleventyComputed:
  pageTitle: {{ home.pageTitle.pt }}
---
```
outputs: {"[object Object]":null} 

I don't know what any other way can I do this? It does look like a very simple scenario. Any help is very appreciated, thanks!

**Comment by pdehaan** [comment]
I might be a little confused here, but...
I have the following ./src/pages/home.liquid page:

```liquid
---
# src/pages/home.liquid
layout: layouts/home.html
eleventyComputed:
  title: "{{ home.pageTitle.pt }}"
---

title=<pre>{{ title }}</pre>
```

And my global ./src/_data/home.js data file:

```js
module.exports = {
  pageTitle: {
    pt: "PT PAGE TITLE",
  },
};
```

And finally a ./src/_includes/layouts/home.html layout file:

```liquid
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>{{ title }}</title>
</head>
<body>

<main>{{ content }}</p>

</body>
</html>
```

The final rendered HTML page looks like this, so I think it's successfully pulling from my global data file via `eleventyComputed` properties in the YAML front matter:

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>PT PAGE TITLE</title>
</head>
<body>

<main>
title=<pre>PT PAGE TITLE</pre>
</p>

</body>
</html>
```

But all of this is hardcoded in a very basic sample repo, so I'm not sure how your CMS is creating files or the layout of your global data files or if there are additional include files of include files, etc. so it might not match your setup exactly. But if you have a public repo somewhere, I can take a look.

**Reply by opexchanger** [reply]
That was a lot of effort from you, I'm really thankful!! That's weird, supposedly it should be working then. I feel like it is tooking too much time out of you, but I made a public clone of the repo: https://github.com/coccamp/public-repo
_layout.html file is in `web/src/_includes/_layout.html` yeah, there's a lot of stuff in there, that's what I'm trying to get rid of 😂
On line 45 I try to access it as `<title>Coccamp | {{ pageTitle }}</title>`, and again testing it on line 224 as 
`<div>{{ eleventyComputed.pageTitle }} e {{ swiper }} e {{ title }}</div>`

The partial where the data should come from is at `web/src/teste.html`. Now it says `pageTitle: '{{ home.aboutCTA }}'` cause I was trying another variables, but to the same end

**Reply by pdehaan** [reply]
I'm not sure how much I can help. 
After cloning and installing the dependencies, I'm getting the following error when trying to build:

```sh
> Error in your Eleventy config file '/private/tmp/public-repo/web/.eleventy.js'.

`EleventyConfigError` was thrown
> Configuration must contain `projectId`

`Error` was thrown:
    Error: Configuration must contain `projectId`
        at exports.initConfig (/private/tmp/public-repo/web/node_modules/@sanity/client/lib/config.js:51:11)
        at SanityClient.config (/private/tmp/public-repo/web/node_modules/@sanity/client/lib/sanityClient.js:78:25)
        at new SanityClient (/private/tmp/public-repo/web/node_modules/@sanity/client/lib/sanityClient.js:46:8)
        at SanityClient (/private/tmp/public-repo/web/node_modules/@sanity/client/lib/sanityClient.js:43:12)
        at Object.<anonymous> (/private/tmp/public-repo/web/src/utils/sanityClient.js:8:18)
```

So looks like the Sanity client needs a specific `projectId` key via .env vars or something, which I don't have. Since it works for you, I guess the next step would be to write some test_home.js script which just tries requiring ./_data/home.js and dumping the contents to the console so you can confirm the data is what you are expecting. Or you could even probably just `console.log()` the values directly from the global data file:


```js
module.exports = async function () {
  const data = await client.fetch(query).catch((err) => console.error(err));
  if (data[0]) {
	console.log({a: data[0]});
    return generateHtml(data[0]);
  }
  console.log({b: data});
  return data;
};
```

I guess I'm also confused why the layout is `layout: layout.html` in https://github.com/coccamp/public-repo/blob/e0edb0eaaa8f830fa46bbe63e17f3a637c6f0189/web/src/teste.html#L2
But then in the repo, the file is `_layout.html` with an underscore at [./web/src/_includes/_layout.html](https://github.com/coccamp/public-repo/blob/main/web/src/_includes/_layout.html). I wasn't sure if something else is touching or renaming/modifying your templates somehow, or if the leading underscore is some other trick I wasn't aware of.

---

Re: [src/teste.html](https://github.com/coccamp/public-repo/blob/main/web/src/teste.html)

```yaml
---
layout: layout.html
title: 'home'
swiper: true
eleventyComputed:
  pageTitle: '{{ home.aboutCTA }}'
---
```

Assuming your global _data/home.js data file ultimately exports an object which has a top-level .aboutCTA property, I would have expected this to work:

```html
<title>Coccamp | {{ pageTitle }}</title>
```

...although since the `pageTitle` is just the global `home.aboutCTA` I wonder if it even needs to be passed around, or if the layout template can just use the `home.aboutCTA` variable directly.


And...

```html
<div>{{ eleventyComputed.pageTitle }} e {{ swiper }} e {{ title }}</div>
```

I wouldn't expect `{{ eleventyComputed.pageTitle }}` to work, since `eleventyComputed` is a special thing and would have created a computed top-level "pageTitle" variable. I'd also expect `{{ swiper }}` (true) and `{{ title }}` ("home") to work since those are declared directly in the template front matter.

It feels like this is a similar concept to my very simple proof of concept, so I'm curious what the fetched contents of the `_data/home.js` are and if it's a Promise and you're missing an `await` somewhere, or if it's an object and if the properties are nested unexpectedly. All of it can get complex pretty quickly when dealing w/ CMS or APIs vs static global data files like I was using.

**Reply by opexchanger** [reply]
Again I'm amazed by the effort you made! Really glad by your help. 
Yeah the thing with the key really is in an .env file, I haven't thought of that. 

The layout thing was a mistake I made when I was trying things to make it work, but apparently eleventy finds the right files in any way, I guess the underscore is more for naming conventions. 

You lead me to some tests I haven't made yet, but for example putting  `<p>{{ home.pageTitle.pt }}</p>` in the _layout.html directly, outputs the expected value alright. 
So basically:
`<p>{{ home.pageTitle.pt }}</p>
      <p>{{ eleventyComputed.pageTitle }}</p>
      <p>{{ pageTitle }}</p>`
outputs only the first p correctly. As as we saw, other hardcoded front-matter values also are passing alright.
The home.js _data file it's working, but the eleventyComputed it's not doing it. 

About your question regarding why not use it directly, it is because there are other pages that will use this same layout. So when we are on page /contact for istance, it will be another eleventyComputed pageTitle value and so on.

As I said I'm really impressed and thankful by the time you took to try to help me! It was really valuable! About the situation, I guess I could try to make a switch or if/else chain on manual front-matter 'pageTitle' variables, and then based on that output the dynamic cms values `{{ home.pageTitle.pt }}` or `{{ contact.pageTitle.pt }}`. This will be ugly but it shoulda work 😂
It really bothers me that I cant get to use this computed tool though, but I guess is what it is lol thanks again bro!!

**Reply by pdehaan** [reply]
Yeah, it’s frustrating. I feel like it should work, and by using global data file directly I think you proved that the syntax and data are correct and in the expected format. Now it’s a case of figuring out how we can create a simple test case that shows the eleventyComputed behavior so we can find a workaround.

Not sure if you can add some `console.log()` statements to the data/home.js data file and paste the output here and I can use that and bypass Sanity client.

**Reply by pdehaan** [reply]
Also, as my brain multitasks between things, are we sure the computed data is correct and not oddly some stringified version of an object? I noticed you have `{{home.aboutCTA}}` in the frontmatter, but `{{home.aboutCTA.pt}}` below for some link text.

```yaml
eleventyComputed:
  pageTitle: '{{ home.aboutCTA }}'
```

```liquid
{{ home.aboutCTA.pt }}
```

We might also need some custom filters to help w/ debugging since it can help debug logging objects.
You could try adding this to your .eleventy.js config file:

```js
eleventyConfig.addFilter("inspect", require("util").inspect);
```

Then see if you get anything interesting from `{{ home.aboutCTA | inspect }}`
